src/Nerdbank.MessagePack/StandardVisitor.cs

```diff
@@ -246,14 +256,75 @@ internal class StandardVisitor(MessagePackSerializer owner) : TypeShapeVisitor
 	/// </summary>
 	/// <typeparam name="T">The data type to make convertible.</typeparam>
 	/// <param name="shape">The type shape.</param>
+	/// <param name="state">An optional state object to pass to the converter.</param>
 	/// <returns>The converter.</returns>
-	protected IMessagePackConverter<T> GetConverter<T>(ITypeShape<T> shape)
+	protected MessagePackConverter<T> GetConverter<T>(ITypeShape<T> shape, object? state = null)
 	{
-		if (owner.TryGetConverter(out IMessagePackConverter<T>? converter))
+		if (owner.TryGetConverter(out MessagePackConverter<T>? converter))
 		{
 			return converter;
 		}
 
-		return this.converters.GetOrAdd<IMessagePackConverter<T>>(shape, this, box => new DelayedConverter<T>(box));
+		return this.converters.GetOrAdd<MessagePackConverter<T>>(shape, this, box => new DelayedConverter<T>(box));
+	}
+
+	/// <summary>
+	/// Gets or creates a converter for the given type shape.
+	/// </summary>
+	/// <param name="shape">The type shape.</param>
+	/// <param name="state">An optional state object to pass to the converter.</param>
+	/// <returns>The converter.</returns>
+	protected IMessagePackConverter GetConverter(ITypeShape shape, object? state = null)
+	{
+		ITypeShapeFunc self = this;
+		return (IMessagePackConverter)shape.Invoke(this, state)!;
+	}
+
+	/// <summary>
+	/// Returns a dictionary of <see cref="MessagePackConverter{T}"/> objects for each subtype, keyed by their alias.
+	/// </summary>
+	/// <param name="objectShape">The shape of the data type that may define derived types that are also allowed for serialization.</param>
+	/// <returns>A dictionary of <see cref="MessagePackConverter{T}"/> objets, keyed by the alias by which they will be identified in the data stream.</returns>
+	/// <exception cref="InvalidOperationException">Thrown if <paramref name="objectShape"/> has any <see cref="KnownSubTypeAttribute"/> that violates rules.</exception>
+	private SubTypes? DiscoverUnionTypes(IObjectTypeShape objectShape)
+	{
+		KnownSubTypeAttribute[]? unionAttributes = objectShape.AttributeProvider?.GetCustomAttributes(typeof(KnownSubTypeAttribute), false).Cast<KnownSubTypeAttribute>().ToArray();
+		if (unionAttributes is null or { Length: 0 })
+		{
+			return null;
+		}
+
+		Dictionary<int, IMessagePackConverter> deserializerData = new();
+		Dictionary<Type, (int Alias, IMessagePackConverter Converter)> serializerData = new();
+		foreach (KnownSubTypeAttribute unionAttribute in unionAttributes)
+		{
+			IObjectTypeShape? subtypeShape = (IObjectTypeShape?)objectShape.Provider.GetShape(unionAttribute.SubType);
```

Comment:
I expect this cast will fail if the subtype happens to implement `IEnumerable`.

Reply:
You are right about that. I've added a test for this and will get this fixed.